pkg/operator/controllers/autosizednodes/autosizednodes_controller.go

```diff
@@ -0,0 +1,125 @@
+package autosizednodes
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/coreos/ignition/v2/config/util"
+	mcv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	"github.com/sirupsen/logrus"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+)
+
+type Reconciler struct {
+	client.Client
+	Scheme *kruntime.Scheme
+
+	log *logrus.Entry
+}
+
+const (
+	configName = "dynamic-node"
+)
+
+func NewReconciler(log *logrus.Entry, mgr ctrl.Manager) *Reconciler {
+	return &Reconciler{
+		Client: mgr.GetClient(),
+
+		log: log,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	var aro arov1alpha1.Cluster
+	var config mcv1.KubeletConfig
+	var err error
+
+	err = r.Get(ctx, request.NamespacedName, &aro)
+	if err != nil {
+		r.log.Error("unable to fetch aro cluster")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	r.log.Infof("Config changed, autoSize: %t\n", aro.Spec.Features.ReconcileAutoSizedNodes)
+
+	// key is used to locate the object in the etcd
+	key := types.NamespacedName{
+		Name: configName,
+	}
+
+	err = r.Get(ctx, key, &config)
+	if client.IgnoreNotFound(err) != nil {
+		err = fmt.Errorf("could not fetch KubeletConfig: %v", err)
+		return ctrl.Result{RequeueAfter: time.Hour}, err
+	}
+
+	if kerrors.IsNotFound(err) {
+		if aro.Spec.Features.ReconcileAutoSizedNodes {
+			//create
+			config := makeConfig()
+
+			err = r.Create(ctx, config, &client.CreateOptions{})
+			err = fmt.Errorf("could not create KubeletConfig: %v", err)
+			return ctrl.Result{}, err
```

Comment:
I think we need "create or update" logic. Currently we do handle the case when the kubelet config already exists, but was borked for any reason (customer error or a bug).

---

## What to watch

I think reconciler should trigger on two things:

1. Change to the `arov1alpha1.Cluster` object with name `arov1alpha1.SingletonClusterName`
2. Change to the `mcv1.KubeletConfig` object with name `configName`.

I beleive we are good here.

## Reconciliation logic

I think the logic should be as follows

### When `aro.Spec.Features.ReconcileAutoSizedNodes` is `true`

1. Get `mcv1.KubeletConfig` with name `configName`
    1. If already exists - update
    1. If `err` == not found - create
    1. If any other error - return an error (will be requeued by controller-runtime automatically)


### When `aro.Spec.Features.ReconcileAutoSizedNodes` is `false`

1. Delete `mcv1.KubeletConfig` with name `configName`
     1. If `err` == not found - do nothing as there is nothing to delete
     1. If any other error - return an error (will be requeued by controller-runtime automatically)